From a9fb14854818ece853944bfd4a327640904055ee Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Wed, 15 May 2019 17:58:06 +0300
Subject: [PATCH 230/386] net: mvpp2: reduce TX FIFO MIN threshold on 2.5G
 ports

2.5G TX FIFO MIN threshold cause CRC errors in TX.
As workaround patch reduce 2.5G ports TX FIFO minimum threshold.

Change-Id: I3e569ca443025a1d629301a2448b6a3e85bcc13e
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/9034
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Marcin Wojtas <marcin@marvell.com>
Reviewed-by: Stefan Chulski <Stefan.Chulski@cavium.com>
[RH: Original patch taken from marvell 88F3720 board support SDK 10.0-PR2003]
Signed-off-by: Ruiqiang Hao <Ruiqiang.Hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2/mvpp2.h    | 10 +++-------
 .../net/ethernet/marvell/mvpp2/mvpp2_main.c   | 19 ++++++-------------
 2 files changed, 9 insertions(+), 20 deletions(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2.h b/drivers/net/ethernet/marvell/mvpp2/mvpp2.h
index 12de906610b2..6e62d2ed2cc5 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2.h
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2.h
@@ -668,14 +668,10 @@
 #define     MVPP2_RX_FC_TRSH_UNIT	256
 
 /* GMAC TX FIFO configuration */
-#define MVPP2_GMAC_TX_FIFO_MIN_TH_1000	\
+#define MVPP2_GMAC_TX_FIFO_MIN_TH	\
 	MVPP2_GMAC_TX_FIFO_MIN_TH_MASK(50)
-#define MVPP2_GMAC_TX_FIFO_MIN_TH_2500	\
-	MVPP2_GMAC_TX_FIFO_MIN_TH_MASK(136)
-#define MVPP2_GMAC_TX_FIFO_LOW_WM_1000		75
-#define MVPP2_GMAC_TX_FIFO_HI_WM_1000		77
-#define MVPP2_GMAC_TX_FIFO_LOW_WM_2500		180
-#define MVPP2_GMAC_TX_FIFO_HI_WM_2500		185
+#define MVPP2_GMAC_TX_FIFO_LOW_WM		75
+#define MVPP2_GMAC_TX_FIFO_HI_WM		77
 
 /* RX buffer constants */
 #define MVPP2_SKB_SHINFO_SIZE \
diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
index 6b9bb066b6b2..3fa59b36f6a1 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
@@ -1815,21 +1815,14 @@ static inline void mvpp2_xlg_max_rx_size_set(struct mvpp2_port *port)
 	writel(val, port->base + MVPP22_XLG_CTRL1_REG);
 }
 
-static void mvpp2_gmac_tx_fifo_configure(struct mvpp2_port *port,
-					 phy_interface_t phy_interface)
+static void mvpp2_gmac_tx_fifo_configure(struct mvpp2_port *port)
 {
 	u32 val, tx_fifo_min_th;
 	u8 low_wm, hi_wm;
 
-	if (phy_interface == PHY_INTERFACE_MODE_2500BASEX) {
-		tx_fifo_min_th = MVPP2_GMAC_TX_FIFO_MIN_TH_2500;
-		low_wm = MVPP2_GMAC_TX_FIFO_LOW_WM_2500;
-		hi_wm = MVPP2_GMAC_TX_FIFO_HI_WM_2500;
-	} else {
-		tx_fifo_min_th = MVPP2_GMAC_TX_FIFO_MIN_TH_1000;
-		low_wm = MVPP2_GMAC_TX_FIFO_LOW_WM_1000;
-		hi_wm = MVPP2_GMAC_TX_FIFO_HI_WM_1000;
-	}
+	tx_fifo_min_th = MVPP2_GMAC_TX_FIFO_MIN_TH;
+	low_wm = MVPP2_GMAC_TX_FIFO_LOW_WM;
+	hi_wm = MVPP2_GMAC_TX_FIFO_HI_WM;
 
 	/* Update TX FIFO MIN Threshold */
 	val = readl(port->base + MVPP2_GMAC_PORT_FIFO_CFG_1_REG);
@@ -1856,7 +1849,7 @@ static void mvpp2_defaults_set(struct mvpp2_port *port)
 	    port->phy_interface == PHY_INTERFACE_MODE_SGMII ||
 	    port->phy_interface == PHY_INTERFACE_MODE_1000BASEX ||
 	    port->phy_interface == PHY_INTERFACE_MODE_2500BASEX)
-		mvpp2_gmac_tx_fifo_configure(port, port->phy_interface);
+		mvpp2_gmac_tx_fifo_configure(port);
 
 	/* Disable Legacy WRR, Disable EJP, Release from reset */
 	tx_port_num = mvpp2_egress_port(port);
@@ -6226,7 +6219,7 @@ static void mvpp2_mac_config(struct net_device *dev, unsigned int mode,
 		mvpp2_xlg_config(port, mode, state);
 	} else {
 		mvpp2_gmac_config(port, mode, state);
-		mvpp2_gmac_tx_fifo_configure(port, state->interface);
+		mvpp2_gmac_tx_fifo_configure(port);
 	}
 
 	if (port->priv->hw_version == MVPP21 && port->flags & MVPP2_F_LOOPBACK)
-- 
2.17.1

